File: src/domain/competence.ts
import { PlainDate } from "./dateptbr";

/**
 * Competencia = mes/ano ao qual a transacao pertence (pela sua propria data),
 * independente da data de emissao do extrato. Correcao C5 da especificacao:
 * um arquivo pode conter varias competencias.
 */
export interface Competence {
  year: number;
  month: number; // 1..12
}

export function competenceOf(date: PlainDate): Competence {
  return { year: date.year, month: date.month };
}

export function competenceKey(c: Competence): string {
  return `${c.year}-${c.month.toString().padStart(2, "0")}`; // "2026-07"
}

export function competenceEquals(a: Competence, b: Competence): boolean {
  return a.year === b.year && a.month === b.month;
}

const MESES_PT = [
  "JANEIRO",
  "FEVEREIRO",
  "MARÇO",
  "ABRIL",
  "MAIO",
  "JUNHO",
  "JULHO",
  "AGOSTO",
  "SETEMBRO",
  "OUTUBRO",
  "NOVEMBRO",
  "DEZEMBRO",
];

/**
 * MonthSheetResolver — mapeia o numero do mes para o nome da aba em PT-BR
 * maiusculo (correcao C1: as abas sao "JANEIRO".."DEZEMBRO", nao "MM-AAAA").
 */
export function monthSheetName(c: Competence): string {
  if (c.month < 1 || c.month > 12) throw new Error(`Mes invalido: ${c.month}`);
  return MESES_PT[c.month - 1];
}

/** Abas de apoio que NUNCA devem receber escrita. */
export const SUPPORT_SHEETS = [
  "Categorias",
  "FLUXO DE CAIXA  SIMPLIFICADO",
  "ANÁLISE",
];

export function isSupportSheet(name: string): boolean {
  return SUPPORT_SHEETS.some(
    (s) => s.replace(/\s+/g, " ").toUpperCase() === name.replace(/\s+/g, " ").toUpperCase(),
  );
}

File: src-tauri/src/domain/competence.rs
//! Competencia = mes/ano ao qual a transacao pertence (pela sua data).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Competence {
    pub year: i32,
    pub month: u8, // 1..=12
}

impl Competence {
    pub fn key(&self) -> String {
        format!("{:04}-{:02}", self.year, self.month)
    }
}

const MESES: [&str; 12] = [
    "JANEIRO", "FEVEREIRO", "MARÇO", "ABRIL", "MAIO", "JUNHO", "JULHO", "AGOSTO", "SETEMBRO",
    "OUTUBRO", "NOVEMBRO", "DEZEMBRO",
];

/// MonthSheetResolver — numero do mes -> nome da aba PT-BR (C1).
pub fn month_sheet_name(c: &Competence) -> &'static str {
    MESES[(c.month as usize) - 1]
}

pub const SUPPORT_SHEETS: [&str; 3] = ["Categorias", "FLUXO DE CAIXA  SIMPLIFICADO", "ANÁLISE"];

pub fn is_support_sheet(name: &str) -> bool {
    SUPPORT_SHEETS
        .iter()
        .any(|s| s.eq_ignore_ascii_case(name.trim()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mapeia_mes_para_aba() {
        assert_eq!(month_sheet_name(&Competence { year: 2026, month: 7 }), "JULHO");
        assert_eq!(month_sheet_name(&Competence { year: 2026, month: 8 }), "AGOSTO");
    }

    #[test]
    fn chave_de_competencia() {
        assert_eq!(Competence { year: 2026, month: 7 }.key(), "2026-07");
    }
}
